### PR TITLE
common/WeightedPriorityQueue: Add new Weighted priority queue

### PR DESCRIPTION
**Pull request by rldleblanc**
This is an update from https://github.com/ceph/ceph/pull/6781 which resolves the issue with the performance degradation with SSDs. Now this queue is showing 3-17% performance improvement across [SSD,HDD] x [4K,4M] tests. The new algorithm is showing very similar enqueue/dequeue times as PrioritizedQueue or a little better. The OP distribution of Weighted Priority stays true in all cases; steady state, contention (more enqueue than dequeue), and relief (more dequeue than enqueue) where the PrioritizedQueue would favor high priority ops and starve all low priority ops during contention. The Weighted Priority Queue also responds well to skewed ops (high priority ops have low costs and low priority ops have high costs). Although, the number of Op deviate from the target dequeue, the costs lines up closely to the target distribution which I believe is one of the primary goals.

It may be possible to increase the performance of this queue by using intrusive containers, but since it performs with the same latency as the current queue and has tangible benefits, I want to get this in sooner than later.

Signed-off-by: Robert LeBlanc <robert.leblanc@endurance.com>

### CODE REVIEW

**Comment by athanatos** [comment]
I still don't want to change the interface to move _strict and _front into an op class parameter.  That behavior is *not* up to the queue implementation, it's required to work the way it does for correctness reasons (especially front vs back).

**Comment by athanatos** [comment]
Every commit needs signoff.

**Comment by athanatos** [comment]
The last few commits need to be merged into the commits which introduced the problems.

**Comment by athanatos** [comment]
"Enable the Weighted PriorityQueue" should probably be two commits:

"osd/: use WeightedPriorityQueue for ShardedOpWQ"
and
"msg/simple/: use WeightedPriorityQueue for DispatchQueue"

**Comment by athanatos** [comment]
Have you tried benchmarking it with both DispatchQueue and ShardedOpWQ using a OpQueue* and making virtual calls?  It may just not make a measurable difference -- that would make it a lot easier to make it configurable at run time.

**Comment by athanatos** [comment]
If you are going to do a style cleanup on PrioritizedQueue, do it as the first commit.

**Comment by dachary** [comment]
@rldleblanc please ignore the false negative from the bot (see http://tracker.ceph.com/issues/14092 for more information). You can get a new run by rebasing & repushing.

File: src/common/WeightedPriorityQueue.h [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+	// Flip a coin to see if this priority gets to run based on cost.
+	// The next op's cost is multiplied by .9 and subtracted from the
+	// max cost seen. Ops with lower costs will have a larger value
+	// and allow them to be selected easier than ops with high costs.
+	if (max_cost == 0 || rand() % max_cost <=
+	    (max_cost - ((dq->second.front().first * 9) / 10))){
+	  break;
+	}
+      }
+      inc_dq();
```
Comment (XinzeChi) [inline src/common/WeightedPriorityQueue.h]: Could you explain why we should inc dq here, maybe this would be more fairness?

Comment (XinzeChi) [inline src/common/WeightedPriorityQueue.h]: Could you explain why we should inc dq here, maybe this would be more fairness?

Comment (rldleblanc) [inline src/common/WeightedPriorityQueue.h]: That is a relic from the round robin code I tried before. It doesn't do anything other than burn CPU now. dq is reassigned in each new pass so it doesn't matter what it is at this point.

**Comment by dachary** [comment]
@rldleblanc It would be useful to do 100 runs of ./run-make-check.sh on this pull request, to assert there are no subtle false negative due to race conditions. The http://jenkins.ceph.dachary.org/job/ceph/LABELS=centos-7&&x86_64/10316/ failure looks like a false negative. The master branch currently has only one known false negative ( http://tracker.ceph.com/issues/14092 ), based on the last ~200 runs or so.

**Comment by XinzeChi** [comment]
@rldleblanc , I do not know why this patch could increase the normal io. Because in normal io, we only have high_queue (which priorty >= 64) and queue which has only item and its priority is 63(from client). But base on your logic, you do not modify the logic when there are high queue and queue which has only item. Do I misread something?

**Comment by rldleblanc** [comment]
In the last commit in OSD.cc, I change which priorities get sent to the
strict queue. I send all client and reops to the normal queue and leave all
the higher (should be zero cost ops in the strict queue). The idea behind
how this improves performance is that there is bound to be an OSD that had
more PGs/objects than others. During times of high load, that OSD can be so
busy handling repops that it will starve normal client IO. This leaves a
few clients with IO that doesn't get serviced because that OSD is to busy.
By sending repops to the weighted priority queue it will allow normal
client ops to go through even in these situations. This causes the clients
who's primary PGS are on other OSDS to be delayed a bit, and they back off
so other clients can get some work done. I've seen that before a few disks
become 100% while some are very idle 30%. With this change almost every
disk is 100% busy all of the time by allowing busy OSDS to get some client
oops done. That is where the performance increase is coming from.

Robert LeBlanc

**Comment by XinzeChi** [comment]
@rldleblanc , thanks for your reply. I think your mean is that it is not so reasonable reops's priority is higher than client's priority at some workload. If that, What about just setting the reops's priority equals with client's priority?

**Comment by rldleblanc** [comment]
No, repops needs to be higher. We have to give them priority otherwise it
will just backup the queues too much and we can get into a deadlock of
sorts. This just prevents client IO from being starved completely for some
clients while others have no blocking. It tries to balance new work with
work that had already been committed by other OSDS.

Robert LeBlanc

File: src/common/WeightedPriorityQueue.h [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+      high_size -= i->second.remove_by_class(k, out);
+      if (i->second.empty()) {
+	high_queue.erase(i++);
+      } else {
+	++i;
+      }
+    }
+  }
+
+  void enqueue_strict(K cl, unsigned priority, T item) override final {
```
Comment (XinzeChi) [inline src/common/WeightedPriorityQueue.h]: I think we could change to parameter `K cl` to `K& cl` and `T item` to `T& item`.
And declare the parameter in OSD::ShardedOpWQ::_enqueue(pair<PGRef, PGQueueable> item) as reference too.

Comment (rldleblanc) [inline src/common/WeightedPriorityQueue.h]: K is just an integer, so I don't think we would buy much, but T may be worthwhile pursuing. Right now, we are just looking to get the new queue in. Later, I was going to spend some more time later trying to optimize it further and look at intrusive lists which from what I understand would just deal with pointers anyways.

**Comment by rldleblanc** [comment]
@dachary I tried running run-make-check.sh on master a week or so ago on Debian Jessie, but it hung up on "PASS: unittest_erasure_code_shec_all". I've pulled the latest master and I'm trying again, but takes a long time to run on my boxes. Would running other (normal) Ceph processes interfere with the test script? I got the same results with my branch so I tested master to make sure I wasn't chasing a ghost.

**Comment by wjwithagen** [comment]
On 5-1-2016 22:28, Robert LeBlanc wrote:
> @dachary <https://github.com/dachary> I tried running run-make-check.sh
> on master a week or so ago on Debian Jessie, but it hung up on "PASS:
> unittest_erasure_code_shec_all". I've pulled the latest master and I'm
> trying again, but takes a long time to run on my boxes. Would running
> other (normal) Ceph processes interfere with the test script? I got the
> same results with my branch so I tested master to make sure I wasn't
> chasing a ghost.

There is some serious performace testing in that set that makes it take
long time.
So for my tests I've made an environmental conditional which I set and
the long tests are skipped.

Part from test/BufferList.cc:
TEST(BufferPtr, copy_out_bench) {
  bool ceph_buffer_no_bench = get_env_bool("CEPH_BUFFER_NO_BENCH");
  for (int s=1; s<=8; s*=2) {
    utime_t start = ceph_clock_now(NULL);
    int buflen = 1048576;
    int count = 1000;
    uint64_t v;
        if (!ceph_buffer_no_bench)
                for (int i=0; i<count; ++i) {
                  bufferptr bp(buflen);
                  for (int64_t j=0; j<buflen; j += s) {
                        bp.copy_out(j, s, (char *)&v);
                  }
                }
    utime_t end = ceph_clock_now(NULL);
    cout << count << " fills of buffer len " << buflen
         << " with " << s << " byte copy_out in "
         << (end - start) << std::endl;
  }
}

Output results are maintained, but fake.
There are more locations where this would be applicable.

Things would perhaps even be better to actually also have a tree with
performance tests. And than merge the perfomance test into there.
And have test mainly for conformace tests.

But that last part is only my 2cts,
--WjW

**Comment by rldleblanc** [comment]
I see output for SKIP, is it possible to just skip performance tests? There
is no way I'm going to meet any performance requirements on a single
spindle.

**Comment by rldleblanc** [comment]
@dachary I've run run-make-check.sh 100 times. Each run has the failure of "FAIL: ceph-detect-init/run-tox.sh tests/test_all.py::TestCephDetectInit::test_debian FAILED", but running on master also had the same failure. There were 11 failures of test/run-rbd-unit-test.sh with 7 showing "./common/RWLock.h: 58: FAILED assert(!is_locked())".

**Comment by rldleblanc** [comment]
@athanatos Testing https://github.com/rldleblanc/ceph/commit/f5120d605372fc76fa5c2d6b6c64f549160dde1f against master does show an improvement in performance, but the standard deviation is generally a bit higher.

It seems that standard deviation is helped by the new queue.

**Comment by rldleblanc** [comment]
@athanatos @liewegas I'm wrapping up the recovery testing, but I have enough data to share this. Under full load, the 4MB recovery test is not able to complete recovery with the Master branch (pulled today). I tested it twice and both times was the same. After about 50% recovery, it just can't make any progress. What is worse yet is that some client I/O is blocked for long times (2847 seconds on the first run after I aborted after 24 minutes of no recovery progress, and 1144 seconds on the second run after aborting after 19 minutes of no progress). With this new queue the recovery completed and the worst times I saw were 278 seconds, but most of the slow I/O messages are < 100 seconds. I'm going to put together the data in some charts for easy consumption once I finish another 4MB test and the 4KB test.

**Comment by markhpc** [comment]
In the recovery tests I've done, I haven't seen recovery ever stall out completely like that, but it can be quite slow.  That's good to know as it might be something we haven't seen before.  Nevertheless, it sounds like the new weighted priority queue is doing great!

**Comment by athanatos** [comment]
It still needs to be made configurable at run time, which means you need to find out whether replacing the instance in the OSD queue and Dispatcher with a unique_ptr to the interface makes a measurable performance difference.

**Comment by liewegas** [comment]
This passed the rados suite (modulo some selinux noise from nrpe).

**Comment by liewegas** [comment]
...except i forgot to specify wip-config on ceph-qa-suite, requeued at http://pulpito.ceph.com/sage-2016-02-02_06:54:17-rados:thrash-wip-sage-testing3---basic-smithi/

**Pull request merged**